content/en/blog/_posts/2018-08-02-dynamically-expand-volume-csi.md

```diff
@@ -0,0 +1,163 @@
+
+Currently, expanding volume is only available with those storage provisioners:
+
+* gcePersistentDisk
+* awsElasticBlockStore
+* OpenStack Cinder
+* glusterfs
+* rbd
+
+In order to enable this feature, we should set feature gate `ExpandPersistentVolumes` true and turn on `PersistentVolumeClaimResize` admission plugin. Once `PersistentVolumeClaimResize` has been enabled, resizing will be allowed by Storage Class whose `allowVolumeExpansion` field is set to true.
+
+Unfortunately, dynamically expanding volume through Container Storage Interface (CSI) and Kubernetes is unavailable, even though the underlying storage providers have this feature.
+
+This article will give a simplified view of CSI, followed by a walkthrough of how to introduce a new expanding volume feature on the existing CSI and Kubernetes. Finally, the article will demonstrate how to dynamically expand volume capacity.
+
+## Container Storage Interface (CSI)
+
+To have a better understanding of what we're going to do, the first thing we need to know is what the Container Storage Interface is. Currently, there are still some problems for already existing storage subsystem within Kubernetes. Storage driver code is maintained in the Kubernetes core repository which is difficult to test. But beyond that, Kubernetes needs to give permissions to storage vendors to check code into the Kubernetes core repository. Ideally, that should be implemented externally.
+
+CSI is designed to define an industry standard that will enable storage providers who enable CSI to be available across container orchestration systems that support CSI.
+
+This diagram depicts a kind of high-level Kubernetes archetypes integrated with CSI:
+
+![csi diagram](/images/blog/2018-08-02-dynamically-expand-volume-csi/csi-diagram.png)
+
+* Three new external components are introduced to decouple Kubernetes and Storage Provider logic;
+* Blue arrows present the conventional way to call against API Server;
+* Red arrows present gRPC to call against Volume Driver.
+
+More details: https://github.com/container-storage-interface/spec/blob/master/spec.md
+
+## Extend CSI and Kubernetes
+
+In order to enable the feature of expanding volume atop Kubernetes, we should extend several components including CSI specification, “in-tree” volume plugin, external-provisioner and external-attacher.
+
+## Extend CSI spec
+
+The feature of expanding volume is still undefined in latest CSI 0.2.0. The new 3 RPCs, including `RequiresFSResize` and `ControllerResizeVolume` and `NodeResizeVolume`, should be introduced.
+
+```
```

Comment:
i know if blackfriday supports the `rpc` code block format.
might as well leave it like that.